server/src/main/java/org/opensearch/search/aggregations/Aggregator.java

```diff
@@ -61,6 +62,8 @@
 @PublicApi(since = "1.0.0")
 public abstract class Aggregator extends BucketCollector implements Releasable {
 
+    final SetOnce<InternalAggregation> internalAggregation = new SetOnce<>();
```

Comment:
Thinking about it more (and taking into account the threading dance between `search` and `search_indexer` thread), we need to have strong boundary established, fe  `ImmutableAggregator` and `MutableAggregator` (current `Aggregator`): so there is no opportunity to call `buildXxx` methods (and alike) from the point where threads handoff  happens.

Reply:
Sorry @reta I don't think I fully understand what you mean here. My understanding of the `Collector` interface is that it's meant to be mutable since Lucene expects a `Collector` object to be passed to the search interface and then returns the same `Collector` object back after having collected the results. I think I'm more or less following the same model in my PR here for aggregations and we try to mitigate calling `buildXXX` after handoff by using the `SetOnce` interface.

Reply:
Tangential to my first point -- I don't think the Lucene concurrent search interface that exists today provides a good way to transform a `Collector` from a `MutableAggregator` to an `ImmutableAggregator` during the handoff either since there's no return value from `search`, we just pass in a collector and let Lucene "mutate" it by gathering the results.

Reply:
Sure @jed326 , I am thinking if we could we separate Aggregator  w/o built `internalAggregation` : the issue as I see it now  - we added the workaround to set the `InternalAggregation` so the `buildXxx` methods would not be called in unexpected thread. However, because we still pass the same `InternalAggregation` over the place, the `buildXxx`  could still be called anywhere (fe by other plugins).

Reply:
Ok @jed326 I think this is not feasible now, how about this: 
 - we remove `setInternalAggregation` completely and replace it with `postCollectionAggregation` (or alike, it is always using the `buildRoot()` and naming is aligned with the phase we expect this method to be called)
 - we remove `getInternalAggregation` and use the same method  `postCollectionAggregation` (which returns value instead)

I think by and large, those are small changes, but they help us to reason about a) what it is being done b) where it is being done c) why it is being done (we would need to add the some comments).

Reply:
Thanks @reta, I was just about to say I think the existing Lucene interface makes it difficult to solve the issue you described in the previous comment.

I was thinking of something sort of similar -- we could remove `setInternalAggregation` and instead store the `InternalAggregation` in `buildTopLevel` itself. This will also guarantee that `buildTopLevel` does not get called multiple times.
For example something like this:
```
public final InternalAggregation buildTopLevel() throws IOException {
        assert parent() == null;
        this.internalAggregation.set(buildAggregations(new long[] { 0 })[0]);
        return internalAggregation.get();
    }
```

With this way we would still keep `getInternalAggregation` and call it during `reduce`. What do you think?

Reply:
I think this is also an option, we just need a very clear indication where the aggregation were being obtained (`postCollectionAggregation()`) from

Reply:
Thanks @reta , I made this change in the latest revision -- let me know what you think!